### PR TITLE
[Fix][Frontend][TOPI] minor bugs

### PR DESCRIPTION
**Pull request by hzfan**
Fix 2 minor bugs in:
- reshape array of shape [1] to []
- pytorch linear

cc @yzhliu @comaniac

### CODE REVIEW

File: python/tvm/relay/frontend/pytorch.py [hunk 1]
```diff
@@ -1445,7 +1445,7 @@ def linear(self, inputs, input_types):
         # 0 - input
         # 1 - weight
         bias = inputs[2]
-        mm_out = self.matmul(inputs[:2], input_types[:2])
+        mm_out = self.matmul([inputs[0], _op.transpose(inputs[1], axes=(1, 0))], input_types[:2])
```
Comment (masahi) [inline python/tvm/relay/frontend/pytorch.py:1448]: We can use `_op.nn.dense` without transposing its second input. Also see https://discuss.tvm.apache.org/t/pytorch-unable-to-import-a-simple-torch-nn-linear-to-relay/10383/4

Comment (comaniac) [reply python/tvm/relay/frontend/pytorch.py:1448]: Maybe I missed something, does that mean the current linear layers converted from PyTorch are wrong?

Comment (masahi) [reply python/tvm/relay/frontend/pytorch.py:1448]: Yes, see my comment in the discuss thread above. Until PT 1.8, `nn.Linear` converts to `aten::addmm`, so this code path never hit unless users explicitly call something that converts to `aten::linear`. Since PT 1.9, `nn.Linear` converts to `aten::linear`, so this becomes a major issue.

Comment (comaniac) [reply python/tvm/relay/frontend/pytorch.py:1448]: Ah I see, so this bug doesn't explode until PT 1.9. Thanks for the explanation!

**Review by junrushao1994 — APPROVED** [review]
I agree with Masa, and otherwise looks good to me :-)

File: python/tvm/relay/frontend/pytorch.py (hunk 1 above)
Comment (hzfan) [reply python/tvm/relay/frontend/pytorch.py:1448]: @masahi  My concern is that `linear` is supposed to support inputs with arbitrary ranks like 1d, 2d, 3d, 4d.... The different input dimensions are normalized in `self.matmul`. If `_op.nn.dense` is directly used, the usability of `linear` would be limited to inputs whose  ranks are 2.

Comment (comaniac) [reply python/tvm/relay/frontend/pytorch.py:1448]: How about we just make an exception: when the input rank is 2 then use dense; otherwise use matmul?

Comment (masahi) [reply python/tvm/relay/frontend/pytorch.py:1448]: Doesn't `op.nn.dense` also support ND input? https://github.com/apache/tvm/blob/850abb0c01afffda071bdeda952f053407005050/python/tvm/relay/op/nn/nn.py#L1528-L1530

Comment (masahi) [reply python/tvm/relay/frontend/pytorch.py:1448]: also if we want to support ND input, please add more tests.

Comment (hzfan) [reply python/tvm/relay/frontend/pytorch.py]: @masahi  Tests with 3d input are added. For now I use `dense` when both inputs are 2d and fallback to `matmul` otherwise.

I tried `op.nn.dense` in ND scenario, but it fails at 
https://github.com/apache/tvm/blob/4b9d43e05b819f1f38420b90abca948070811687/python/tvm/topi/x86/dense_alter_op.py#L29-L37

where `data_tensor.shape` is unpacked into a tuple of size 2. I also check that in `tests/python/relay/test_op_level1.py` `dense` is tested numerically for 2d inputs only.  Not sure if I miss anything.

**Review by comaniac — APPROVED** [review]

File: python/tvm/relay/frontend/pytorch.py (hunk 1 above)
Comment (masahi) [reply python/tvm/relay/frontend/pytorch.py]: Yeah it looks like relay doc is wrong and topi dense indeed only supports 2D https://github.com/apache/tvm/issues/8412

**Comment by hzfan** [comment]
Any idea why ci always gets aborted in Build:CPU? Did I miss anything?

**Comment by comaniac** [comment]
> Any idea why ci always gets aborted in Build:CPU? Did I miss anything?

It was timeout (4hrs). Re-trigger CI may resolve this issue

**Comment by masahi** [comment]
It should be fixed by https://github.com/apache/tvm/pull/8658

**Comment by junrushao1994** [comment]
Thanks @hzfan @masahi @comaniac! It is merged :-)
